File: apps/runtime/src/git/repository/branches.rs
//! Branch listing plus the reference lookups every other operation resolves
//! its arguments through.

use super::*;

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BranchRecord {
    pub name: String,
    pub full_ref: String,
    pub oid: String,
    pub remote: bool,
    pub current: bool,
    pub upstream: Option<String>,
    pub ahead: Option<u64>,
    pub behind: Option<u64>,
    pub upstream_missing: bool,
    pub symbolic_target: Option<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct BranchSnapshot {
    pub repository_id: String,
    pub repository_path: String,
    pub head: ExpectedState,
    pub branches: Vec<BranchRecord>,
    pub remotes: Vec<String>,
    pub observed_at: String,
}

impl RepositoryService {
    pub async fn branches(
        &self,
        workspace_root: &Path,
        requested: &str,
    ) -> AppResult<BranchSnapshot> {
        let context = self.context(workspace_root, requested).await?;
        let token = Cancellation::default();
        let head = self.head(&context.repository, &token).await?;
        let output = self.read(&context.repository, args(&["for-each-ref", "--sort=refname", "--format=%(refname)%00%(objectname)%00%(upstream)%00%(upstream:track,nobracket)%00%(symref)%00", "refs/heads/", "refs/remotes/"]), &token).await?;
        let mut branches = Vec::new();
        for record in fields_with_lf(&output, 5)? {
            let full_ref = record[0].clone();
            let remote = full_ref.starts_with("refs/remotes/");
            let name = full_ref
                .strip_prefix(if remote {
                    "refs/remotes/"
                } else {
                    "refs/heads/"
                })
                .ok_or_else(malformed)?
                .to_owned();
            if !valid_oid(&record[1]) {
                return Err(malformed());
            }
            let (ahead, behind, missing) = parse_tracking(&record[3])?;
            branches.push(BranchRecord {
                current: !remote && head.branch.as_ref() == Some(&name),
                name,
                full_ref,
                oid: record[1].clone(),
                remote,
                upstream: nonempty(&record[2]),
                ahead: if record[2].is_empty() { None } else { ahead },
                behind: if record[2].is_empty() { None } else { behind },
                upstream_missing: missing,
                symbolic_target: nonempty(&record[4]),
            });
        }
        Ok(BranchSnapshot {
            repository_id: context.repository_id(),
            repository_path: path_string(&context.repository)?,
            head,
            branches,
            remotes: self.remotes(&context.repository, &token).await?,
            observed_at: now(),
        })
    }

    pub(super) async fn head(
        &self,
        directory: &Path,
        token: &Cancellation,
    ) -> AppResult<ExpectedState> {
        let branch_output = self
            .output(
                directory,
                args(&["symbolic-ref", "--quiet", "--short", "HEAD"]),
                self.command_timeout.min(Duration::from_secs(15)),
                token,
                None,
            )
            .await?;
        let branch = match branch_output.status {
            Some(0) => Some(one_line(&branch_output.stdout)?.to_owned()),
            Some(1) => None,
            _ => return Err(command_error(&branch_output)),
        };
        let output = self
            .output(
                directory,
                args(&["rev-parse", "--verify", "--quiet", "HEAD^{commit}"]),
                self.command_timeout.min(Duration::from_secs(15)),
                token,
                None,
            )
            .await?;
        let head_oid = if output.status == Some(0) {
            let oid = one_line(&output.stdout)?;
            if !valid_oid(oid) {
                return Err(malformed());
            }
            Some(oid.to_owned())
        } else if output.status == Some(1) && branch.is_some() {
            None
        } else {
            return Err(command_error(&output));
        };
        Ok(ExpectedState { head_oid, branch })
    }

    pub(super) async fn resolve(
        &self,
        directory: &Path,
        reference: &str,
        token: &Cancellation,
    ) -> AppResult<String> {
        self.validate_reference(directory, reference, token).await?;
        let output = self
            .read(
                directory,
                vec![
                    "rev-parse".into(),
                    "--verify".into(),
                    "--end-of-options".into(),
                    format!("{reference}^{{commit}}"),
                ],
                token,
            )
            .await?;
        let oid = one_line(&output)?;
        if !valid_oid(oid) {
            return Err(malformed());
        }
        Ok(oid.to_owned())
    }

    pub(super) async fn validate_branch(
        &self,
        directory: &Path,
        name: &str,
        token: &Cancellation,
    ) -> AppResult<()> {
        if name.is_empty()
            || name.len() > 255
            || name.starts_with('-')
            || name.starts_with("refs/")
            || name.contains("@{")
            || name.chars().any(char::is_control)
        {
            return Err(AppError::BadRequest("Branch name is invalid".into()));
        }
        self.read(
            directory,
            vec!["check-ref-format".into(), "--branch".into(), name.into()],
            token,
        )
        .await
        .map(|_| ())
        .map_err(|_| AppError::BadRequest("Branch name is invalid".into()))
    }

    pub(super) async fn validate_reference(
        &self,
        directory: &Path,
        reference: &str,
        token: &Cancellation,
    ) -> AppResult<()> {
        if reference == "HEAD" || valid_oid(reference) {
            return Ok(());
        }
        if reference.starts_with("refs/") {
            if reference.len() > 1024 || reference.chars().any(char::is_control) {
                return Err(AppError::BadRequest("Git reference is invalid".into()));
            }
            self.read(
                directory,
                vec!["check-ref-format".into(), reference.into()],
                token,
            )
            .await
            .map(|_| ())
            .map_err(|_| AppError::BadRequest("Git reference is invalid".into()))
        } else {
            self.validate_branch(directory, reference, token).await
        }
    }

    pub(super) async fn commit_refs(
        &self,
        directory: &Path,
        token: &Cancellation,
    ) -> AppResult<HashMap<String, Vec<String>>> {
        let output = self
            .read(
                directory,
                args(&[
                    "for-each-ref",
                    "--format=%(objectname)%00%(*objectname)%00%(refname)%00",
                    "refs/heads/",
                    "refs/remotes/",
                    "refs/tags/",
                ]),
                token,
            )
            .await?;
        let mut refs: HashMap<String, Vec<String>> = HashMap::new();
        for record in fields_with_lf(&output, 3)? {
            let oid = if record[1].is_empty() {
                &record[0]
            } else {
                &record[1]
            };
            refs.entry(oid.clone()).or_default().push(record[2].clone());
        }
        Ok(refs)
    }
}

File: apps/desktop/src/core/git/repository/branches.ts
import {
  badRequest,
  malformed,
  nonempty,
  nowRfc3339,
  validOid,
} from "../support";
import { fieldsWithLf, parseTracking } from "./parse";
import { RepositoryService, repositoryId } from "./service";
import type {
  BranchRecord,
  BranchSnapshot,
  IdentityRecord,
  RemoteRecord,
  TagRecord,
  TagSnapshot,
} from "./types";
import { sanitizeRepository } from "../support";

/**
 * The three ref namespaces one checkout answers about: branches, tags and
 * remotes, plus the identity a commit from it would carry.
 *
 * Ports `branches.rs`, `refs.rs` and `identity.rs`. They share a file because
 * all four reads are one `for-each-ref` or one `git config` and nothing else,
 * and because every one of them is shaped by the same decision: an absent
 * value is `null`, never zero and never an error. A branch without an upstream
 * reports `null` ahead/behind because "nothing to push" and "nowhere to push"
 * are different answers, and a machine with no `user.email` is a normal machine
 * that has simply not committed yet.
 */

const MAX_TAGS = 5_000;
const MAX_REMOTES = 256;

export async function branches(
  service: RepositoryService,
  workspaceRoot: string,
  requested: string,
): Promise<BranchSnapshot> {
  const context = await service.context(workspaceRoot, requested);
  const head = await service.head(context.repository);
  const output = await service.read(context.repository, [
    "for-each-ref",
    "--sort=refname",
    "--format=%(refname)%00%(objectname)%00%(upstream)%00%(upstream:track,nobracket)%00%(symref)%00",
    "refs/heads/",
    "refs/remotes/",
  ]);
  const records: BranchRecord[] = [];
  for (const record of fieldsWithLf(output, 5)) {
    const fullRef = record[0] as string;
    const remote = fullRef.startsWith("refs/remotes/");
    const prefix = remote ? "refs/remotes/" : "refs/heads/";
    if (!fullRef.startsWith(prefix)) throw malformed();
    const name = fullRef.slice(prefix.length);
    if (!validOid(record[1] as string)) throw malformed();
    const [ahead, behind, missing] = parseTracking(record[3] as string);
    const tracked = record[2] !== "";
    records.push({
      current: !remote && head.branch === name,
      name,
      fullRef,
      oid: record[1] as string,
      remote,
      upstream: nonempty(record[2] as string),
      ahead: tracked ? ahead : null,
      behind: tracked ? behind : null,
      upstreamMissing: missing,
      symbolicTarget: nonempty(record[4] as string),
    });
  }
  return {
    repositoryId: repositoryId(context),
    repositoryPath: context.repository,
    head,
    branches: records,
    remotes: await service.remotes(context.repository),
    observedAt: nowRfc3339(),
  };
}

const TAG_FORMAT =
  "--format=%(refname)%00%(objectname)%00%(objecttype)%00%(*objectname)%00%(taggername)%00%(taggerdate:iso-strict)%00%(contents:subject)%00";

export async function tags(
  service: RepositoryService,
  workspaceRoot: string,
  requested: string,
): Promise<TagSnapshot> {
  const context = await service.context(workspaceRoot, requested);
  const head = await service.head(context.repository);
  const output = await service.read(context.repository, [
    "for-each-ref",
    "--sort=-creatordate",
    TAG_FORMAT,
    "refs/tags/",
  ]);
  const records = fieldsWithLf(output, 7);
  if (records.length > MAX_TAGS) {
    throw badRequest("This repository has more tags than this view supports");
  }
  return {
    repositoryId: repositoryId(context),
    repositoryPath: context.repository,
    head,
    tags: records.map((record) => tagFromRecord(record)),
    observedAt: nowRfc3339(),
  };
}

function tagFromRecord(record: string[]): TagRecord {
  const fullRef = record[0] as string;
  if (!fullRef.startsWith("refs/tags/")) throw malformed();
  const name = fullRef.slice("refs/tags/".length);
  // A lightweight tag has no dereferenced object, so the ref's own object is
  // already the commit.
  const targetOid =
    record[3] === "" ? (record[1] as string) : (record[3] as string);
  if (!validOid(record[1] as string) || !validOid(targetOid)) throw malformed();
  return {
    name,
    fullRef,
    oid: record[1] as string,
    targetOid,
    annotated: record[2] === "tag",
    subject: nonempty(record[6] as string),
    taggerName: nonempty(record[4] as string),
    taggerTime: nonempty(record[5] as string),
  };
}

/** One tag by name, or `undefined` when nothing holds that name. */
export async function tagRecord(
  service: RepositoryService,
  repository: string,
  name: string,
  signal?: AbortSignal,
): Promise<TagRecord | undefined> {
  await service.validateTagName(repository, name, signal);
  const output = await service.read(
    repository,
    ["for-each-ref", TAG_FORMAT, `refs/tags/${name}`],
    signal,
  );
  const records = fieldsWithLf(output, 7);
  const record = records[0];
  if (record === undefined) return undefined;
  if (records.length !== 1 || record[0] !== `refs/tags/${name}`) {
    throw malformed();
  }
  return { ...tagFromRecord(record), name };
}

/**
 * Remote URLs are redacted before they leave the service; Armadra stores none
 * of them, and a redacted value must not be sent back as an update.
 */
export async function remoteRecords(
  service: RepositoryService,
  workspaceRoot: string,
  requested: string,
): Promise<RemoteRecord[]> {
  const context = await service.context(workspaceRoot, requested);
  const names = await service.remotes(context.repository);
  if (names.length > MAX_REMOTES) {
    throw badRequest(
      "This repository has more remotes than this view supports",
    );
  }
  const records: RemoteRecord[] = [];
  for (const name of names) {
    const fetch = (
      await service.read(context.repository, ["remote", "get-url", "--", name])
    )
      .toString("utf8")
      .replace(/\n$/, "");
    const push = (
      await service.read(context.repository, [
        "remote",
        "get-url",
        "--push",
        "--",
        name,
      ])
    )
      .toString("utf8")
      .replace(/\n$/, "");
    const fetchUrl = sanitizeRepository(fetch);
    const pushUrl = sanitizeRepository(push);
    records.push({
      name,
      fetchUrl,
      pushUrl,
      redacted: fetchUrl !== fetch || pushUrl !== push,
    });
  }
  return records;
}

/**
 * `git config user.name` / `user.email` for one checkout.
 *
 * A plain config read, so it needs no execution grant: nothing about it runs a
 * repository filter. Unset is `null`, never an error.
 */
export async function identity(
  service: RepositoryService,
  workspaceRoot: string,
  requested: string,
): Promise<IdentityRecord> {
  const context = await service.context(workspaceRoot, requested);
  return {
    name: await configValue(service, context.repository, "user.name"),
    email: await configValue(service, context.repository, "user.email"),
  };
}

/**
 * One config value, or `null` when it is unset.
 *
 * `git config --get` exits 1 for "no such key", which is an answer rather than
 * a failure. `--end-of-options` keeps a key beginning with a dash from being
 * read as a flag.
 */
async function configValue(
  service: RepositoryService,
  directory: string,
  key: string,
): Promise<string | null> {
  const output = await service.output(
    directory,
    ["config", "--get", "--end-of-options", key],
    Math.min(service.commandTimeoutMs, 10_000),
  );
  if (output.status !== 0) return null;
  // A key configured to the empty string is as unset as an absent one for the
  // two things this answer is used for.
  return nonempty(output.stdout.toString("utf8").replace(/\n$/, "").trim());
}
